Issue 31: Implement tests for Auth/Account APIs

Nevedrov opened the pull request:
feat: create Auth and Account API controller test integrated with allure

Nevedrov committed:
feat: add AuthController tests

Nevedrov committed:
chore: add LoginResponse and RefreshResponse json  validation schemas for \auth\login and \auth\refresh tests

Nevedrov committed:
build: update pom.xml

Nevedrov committed:
fix: fixed auth tests, reworked tests with mocMmvc

Nevedrov committed:
chore: add test.application.properties configuration file

Nevedrov committed:
chore: update json validation schemas, fix "$schema" field url

Nevedrov committed:
refactor: split AuthControllerTests setup code into separate helper classes
fix: fixed response body fields incorrect naming

Nevedrov committed:
feat: add Accounts API tests

Nevedrov committed:
chore: add Account API tests json validation schemas

Nevedrov committed:
build: update pom file

Nevedrov committed:
chore: update test.application.properties

Nevedrov committed:
refactor: remove unused imports

Nevedrov committed:
chore: delete unnecessary schemas dir

Nevedrov committed:
chore: delete unnecessary file

Nevedrov committed:
chore: hide test.application.properties server setup properties

Nevedrov committed:
refactor:
- move MockMvc field to RequestExecutor class
- add request and response attachments
- remove print of request and response in stdout

Nevedrov committed:
fix:
- moved "Do not edit" comment from server.url property string

Nevedrov committed:
feat:
- add AttachmentBuilder and SuiteHierarchyProvider helper classes

Nevedrov committed:
chore:
- create allure.properties file
- create allure-report-table.css for pretty attachment html documents view
- add aop-ajc.xml for AspectJ weaver proper work

Nevedrov committed:
build:
- update pom dependencies

Nevedrov committed:
refactor:
- change LoginProvider method names and call sequence
- add Allure steps annotations

Nevedrov committed:
refactor:
- change AuthProvider method signatures
- remove unnecessary authorization flag parameter in Logout method

Nevedrov committed:
refactor:
- change AccountProvider method signatures

Nevedrov committed:
fix:
- fix hibernate.StaleObjectStateException by removing user.setId method call
- userRepo clearing all users before each creation

Nevedrov committed:
refactor:
- move tests from unnecessary FirstLoginTests nested classes
- break initializeMocks setup method into separate methods. Tests initialization method renamed to Setup
- update AuthControllerTests requests wrapper helper functions signatures. Annotate steps
- Set test info for Allure reports

Nevedrov committed:
refactor:
- move tests from unnecessary FirstLoginTests nested classes
- break initializeMocks setup method into separate methods. Tests initialization method renamed to Setup
- update UserControllerTests requests wrapper helper functions signatures. Annotate steps
- Set test info for Allure reports

Nevedrov committed:
fix:
- fix account/change-password status code in description and display name
- remove unnecessary "TestLogoutWithoutLogin" test in AuthControllerTests
- fix invalid Allure report view suite hierarchy by removing Allure suite setter from global SuiteHierarchyProvider to into setup method of controllers tests classes

PavloKilko commented on kub-backend-ce/src/test/resources/schemas/UserDetailsResponse.json:
```diff
@@ -0,0 +1,40 @@
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "title": "UserChangePasswordResponse",
```
Сhange on UserDetailsResponse

Nevedrov replied:
Done

Nevedrov committed:
chore:
- change UserDetailsResponse schema title field

PavloKilko reviewed (approved):
lgtm

Nevedrov committed:
build: update allure-maven plugin reportVersion to fix "UnrecognizedPropertyException" while generating allure report via allure-maven plugin

YaroslavDrovolskyi commented on kub-backend-ce/src/test/java/education/kub/backend/ce/domain/user/controller/UserControllerTests.java:
```diff
@@ -0,0 +1,282 @@
+            ValidateChangePasswordResponse(Map.of("new_password", new_password), HttpStatus.UNAUTHORIZED);
+        }
+
+        @Test
+        @DisplayName("When request body without new password, /api/v1/account/change-password returns 401")
+        @Description("When request body without new password, /api/v1/account/change-password returns 401.")
+        public void ChangePasswordWithoutNewPassword() {
+            SetAllureTestSubSuite();
+            FirstLogin();
+            ValidateChangePasswordResponse(Map.of("password", password), HttpStatus.BAD_REQUEST);
+        }
```
either description (`401`), either code (`BAD_REQUEST`) is incorrect

Nevedrov replied:
Fixed.

YaroslavDrovolskyi commented on kub-backend-ce/src/test/java/education/kub/backend/ce/domain/user/controller/UserControllerTests.java:
```diff
@@ -0,0 +1,282 @@
+            ValidateChangePasswordResponse(Map.of("password", password), HttpStatus.BAD_REQUEST);
+        }
+
+        @Test
+        @DisplayName("When request body with empty new password, /api/v1/account/change-password returns 401")
+        @Description("When request body with empty new password, /api/v1/account/change-password returns 401.")
+        public void ChangePasswordWithEmptyNewPassword() {
+            SetAllureTestSubSuite();
+            FirstLogin();
+            ValidateChangePasswordResponse(Map.of("old_password", password, "new_password", ""), HttpStatus.UNPROCESSABLE_ENTITY);
+        }
```
either description (`401`), either code (`UNPROCESSABLE_ENTITY`) is incorrect

Nevedrov replied:
Fixed.

YaroslavDrovolskyi commented on kub-backend-ce/src/test/java/education/kub/backend/ce/domain/user/controller/UserControllerTests.java:
```diff
@@ -0,0 +1,282 @@
+            ValidateRecoverPasswordResponse(Map.of(), HttpStatus.BAD_REQUEST);
+        }
+
+        @Test
+        @DisplayName("When request body with empty email, /api/v1/account/recovery-password returns 401")
+        @Description("When request body with empty email, /api/v1/account/recovery-password returns 401.")
+        public void RecoverPasswordWithEmptyEmail() {
+            SetAllureTestSubSuite();
+            ValidateRecoverPasswordResponse(Map.of("email", ""), HttpStatus.UNPROCESSABLE_ENTITY);
+        }
+
+        @Test
+        @DisplayName("When request body with invalid email, /api/v1/account/recovery-password returns 401")
+        @Description("When request body with invalid email, /api/v1/account/recovery-password returns 401.")
+        public void RecoverPasswordWithInvalidEmail() {
+            SetAllureTestSubSuite();
+            ValidateRecoverPasswordResponse(Map.of("email", "affafdad"), HttpStatus.UNPROCESSABLE_ENTITY);
+        }
```
either description (`401`), either code (`UNPROCESSABLE_ENTITY`) is incorrect

Nevedrov replied:
Fixed.

YaroslavDrovolskyi commented on kub-backend-ce/src/test/java/education/kub/backend/ce/helpers/users/UserProvider.java:
```diff
@@ -0,0 +1,54 @@
+        user.setStatus(ACTIVATED);
+        user.setTemporaryPasswordHashed(null);
+        user.setTemporaryPasswordExpiresAt(null);
+        user.setCreatedAt(null);
```
maybe, `createdAt` field is automatically handled by JPA?

Nevedrov replied:
Removed all unnecessary null assignments.

Nevedrov committed:
fix: AttachmentBuilder now returns empty body cell if body json string in response is empty

Nevedrov committed:
refactor:
- remove unnecessary user entity assignments in UserProvider
- add component annotation

Nevedrov committed:
fix:
- remove unnecessary AutoConfigureMockMvc annotation
- fix ChangePasswordWithoutNewPassword test old_password request body field name
- fix several test title and description status code

Nevedrov committed:
refactor:
- remove unnecessary AutoConfigureMockMvc annotation
